### PR TITLE
Added more functionality, especially math related.

### PR DESCRIPTION
**Pull request by TechnichalDoge**
Coded in a lot of functions for trigonometry to be used on vector2s, both those that use them as direct inputs instead of two different values ( atan(y,x) => atan(vec2(y,x)) ) for organisation reasons, as well as the ones that manipulate both values of a vector 2 and outputs a vector 2.
Added a plusMinus function that outputs a vector two of the results of addition and subtraction (for square roots and projectile motion, and a lot more) This was the reason for the above.
Added the start of a trigonometry/geometry section, as well as added polar2dtogrid2d and polar3dtogrid3d, which would offload some of the math into glua.
Of cource added all the descriptions.

This is one of my first times working with lua/glua so please be kind and please try to fix it, if you want.

### CODE REVIEW

File: lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua [hunk 1]
```diff
@@ -52,4 +52,21 @@ e2function void entity:ejectPodTo(vector pos)
...
     driver:ExitVehicle()
     driver:SetPos(clampedPos)
-end
\ No newline at end of file
+end
+
+
+-- Easy way to have some of the setPos benefits without mostly being able to be abused.
+e2function number entity:setFairPos(vector pos)
+    if this:IsPlayer() then return 3 end -- Debug output 3 if the target is a player, because we dont want someone tping themselves around the e2 or any other player.
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: Typically you will simply return `false` here. If you needed the user to know what went wrong you could add a ChatPrint in here. Numeric return codes aren't useful here.

Comment is unnecessary as it's self explanatory.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua [hunk 2]
```diff
@@ -52,4 +52,21 @@ e2function void entity:ejectPodTo(vector pos)
 
     driver:ExitVehicle()
     driver:SetPos(clampedPos)
-end
\ No newline at end of file
+end
+
+
+-- Easy way to have some of the setPos benefits without mostly being able to be abused.
+e2function number entity:setFairPos(vector pos)
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: `pos` is very vague, perhaps `desiredPos`, or `givenPos` would make the code easier to read.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua [hunk 3]
```diff
@@ -52,4 +52,21 @@ e2function void entity:ejectPodTo(vector pos)
 
     driver:ExitVehicle()
     driver:SetPos(clampedPos)
-end
\ No newline at end of file
+end
+
+
+-- Easy way to have some of the setPos benefits without mostly being able to be abused.
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: The name of the function should probably explain what it does well enough. If there's any uncertainty, the E2 docs should clear it up. Users need to know what this function does, we should already know without needing a comment here.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua [hunk 4]
```diff
@@ -52,4 +52,21 @@ e2function void entity:ejectPodTo(vector pos)
...
+e2function number entity:setFairPos(vector pos)
+    if this:IsPlayer() then return 3 end -- Debug output 3 if the target is a player, because we dont want someone tping themselves around the e2 or any other player.
+    local posToLoc = pos - self:GetPos() -- distance to the wanted position
+    local lengthPos = posToLoc[1]^2 + posToLoc[2]^2 + posToLoc[3]^2 -- we dont need to do the square root on this, itll just waste time and resources
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: Of all of the things you leave comments on, you neglected to explain what this math actually _does_. The name of the variable should explain what the result of the math is, and the comment should explain the methodology behind getting that number.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua [hunk 5]
```diff
@@ -52,4 +52,21 @@ e2function void entity:ejectPodTo(vector pos)
...
+
+    local entB4PosLoc = this:GetPos() - self:GetPos() -- get distance to target entity
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: I think this variable name should be changed. It's quite cryptic.

If you intend to write "before", just write "before".

File: lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua [hunk 6]
```diff
@@ -52,4 +52,21 @@ e2function void entity:ejectPodTo(vector pos)
...
+    local entLength = entB4PosLoc[1]^2 + entB4PosLoc[2]^2 + entB4PosLoc[3]^2 -- we dont need to do the square root on this, itll just waste time and resources
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: Again, this could use a better name and a comment explaining why this math is being done.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_miscellaneous.lua [hunk 7]
```diff
@@ -0,0 +1,5 @@
+E2Lib.RegisterExtension( "cfc_e2_lib", true )
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_miscellaneous.lua]: We should have a file dedicated to constants if that's what we want to do. It should be scoped for `/custom/constants/cfc_contants.lua` or /custom/cfc_constants.lua`

File: lua/entities/gmod_wire_expression2/core/custom/cfc_miscellaneous.lua [hunk 8]
```diff
@@ -0,0 +1,5 @@
...
+
+-- Boolean Constants (we make it easy so you dont have to)
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_miscellaneous.lua]: This should be a separate PR of its own with its own discussion.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua [hunk 9]
```diff
@@ -52,4 +52,21 @@ e2function void entity:ejectPodTo(vector pos)
...
+
+    if lengthPos >= CFCMaxFairSetDist^2 then return 1 end -- Debug output 1 if the wanted pos is out of bounds and end
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: Assign `CFCMaxFairSetDist^2` to a descriptive variable to do the comparison.

Also, no need to return anything here. `nil` or `false` is sufficient.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua [hunk 10]
```diff
@@ -52,4 +52,21 @@ e2function void entity:ejectPodTo(vector pos)
...
+    if entLength >= CFCMaxFairSetDist^2 then return 2 end -- Debug output 2 if the target entity pos is out of bounds and end
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: Same comment as before. `CFCMaxFairSetDist^2` should be its own var so the comparison is self-explanatory.

`nil` or `false` is a sufficient return here.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua [hunk 11]
```diff
@@ -52,4 +52,21 @@ e2function void entity:ejectPodTo(vector pos)
...
+
+    return 0 --passed all tests, debug 0 for success and move on
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: You can't return before you actually set the position. This is a no-op function right now.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_number.lua [hunk 12]
```diff
@@ -0,0 +1,17 @@
+E2Lib.RegisterExtension( "cfc_e2_lib", true )
+
+__e2setcost(2)
+
+e2function vector2 number:plusMinus(number otherNum)
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_number.lua]: This is a very nondescript function. I also struggle to find it's usefulness

File: lua/entities/gmod_wire_expression2/core/custom/cfc_number.lua [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+__e2setcost(2)
+
+e2function vector2 number:plusMinus(number otherNum)
+    return {
+            this + otherNum,
+            this - otherNum
+    }
+end
+
+e2function vector2 plusMinus(number firstNum, number otherNum)
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_number.lua]: Same comment as before

File: lua/entities/gmod_wire_expression2/core/custom/cfc_number.lua [hunk 14]
```diff
@@ -0,0 +1,17 @@
+E2Lib.RegisterExtension( "cfc_e2_lib", true )
+
+__e2setcost(2)
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_number.lua]: Do these simple operations need a cost?

File: lua/entities/gmod_wire_expression2/core/custom/cfc_trigonometry_geometry.lua [hunk 15]
```diff
@@ -0,0 +1,38 @@
+E2Lib.RegisterExtension( "cfc_e2_lib", true )
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_trigonometry_geometry.lua]: This file needs to be its own PR. I won't review it in here.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_vector2_NumberOperands.lua [hunk 16]
```diff
@@ -0,0 +1,396 @@
+E2Lib.RegisterExtension( "cfc_e2_lib", true )
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cfc_vector2_NumberOperands.lua]: This file needs to be its own PR. I won't review it in here.

Comment (brandonsturgeon) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_vector2_NumberOperands.lua]: This file also needs to be converted to spaces.

File: lua/entities/gmod_wire_expression2/core/custom/cl_cfc_e2_lib.lua [hunk 17]
```diff
@@ -1,46 +1,103 @@
 -- PvP Functions
```
Comment (brandonsturgeon) [inline lua/entities/gmod_wire_expression2/core/custom/cl_cfc_e2_lib.lua]: This file will need to be split up as the other files are pulled out into their own PR.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_convars.lua [hunk 18]
```diff
@@ -0,0 +1,3 @@
+E2Lib.RegisterExtension( "cfc_e2_lib", true )
+
+CFCMaxFairSetDist = CreateConVar("CFC_Lib_MaxFairSetDistance", 250)
```
[the remark this answers is not in the document]
Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_convars.lua]: I like things customizabile in way to make it easy.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_array.lua [hunk 19]
```diff
@@ -1,18 +1,19 @@
 E2Lib.RegisterExtension( "cfc_e2_lib", true )
 
+-- Shuffle the indexes in a given array
```
[the remark this answers is not in the document]
Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_array.lua]: Pressed 'commit suggestion'

File: lua/entities/gmod_wire_expression2/core/custom/cfc_convars.lua (hunk 18 above)
[the remark this answers is not in the document]
Comment (brandonsturgeon) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_convars.lua]: Right but does it need to be assigned to a variable? I thought you could just set the convar

[the remark this answers is not in the document]
Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_convars.lua]: Done

File: lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua (hunk 3 above)
Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: Removed the comment

File: lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua (hunk 2 above)
Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: changed it to target

File: lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua (hunk 1 above)
Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: i see an use case of people using the debug codes for other things as well.

Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: Removed comment

Comment (brandonsturgeon) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: There may be such a case, but this is an antipattern for E2. Very few, if any other functions do such a thing.

Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: Could i make it a client convar or something

Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_entity.lua]: I agree with you, no other functions do debug like this, and so i will remove them.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_trigonometry_geometry.lua (hunk 15 above)
Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_trigonometry_geometry.lua]: How do i split the pr up?

File: lua/entities/gmod_wire_expression2/core/custom/cfc_vector2_NumberOperands.lua [hunk 20]
```diff
@@ -0,0 +1,396 @@
...
+e2function vector2 abs(vector2)
+	return {
+		abs(vector2[1]),
+		abs(vector2[2])
+	}
+end
+
+--- rounds towards zero
+e2function vector2 int(vector2)
+	local function intNum(number)
```
[the remark this answers is not in the document]
Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_vector2_NumberOperands.lua]: Did not see round, floor, nor ceil while looking through the code, thanks! As for the function inside the function case, i might just replace it by just modifying the parts the function is being used on.

File: lua/entities/gmod_wire_expression2/core/custom/cfc_trigonometry_geometry.lua (hunk 15 above)
Comment (PLally) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_trigonometry_geometry.lua]: create a new branch and pr with this file and remove it from this branch

File: lua/entities/gmod_wire_expression2/core/custom/cfc_vector2_NumberOperands.lua [hunk 21]
```diff
@@ -0,0 +1,396 @@
...
+
+
+e2function vector2 toStringV2(vector2 vector)
+	return {
+    	tostring(vector[1]),
+    	tostring(vector[2])
+  	}
+end
+
+e2function vector2 vector2:toStringV2()
```
[the remark this answers is not in the document]
Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_vector2_NumberOperands.lua]: These functions already exist anyway

File: lua/entities/gmod_wire_expression2/core/custom/cfc_trigonometry_geometry.lua (hunk 15 above)
Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_trigonometry_geometry.lua]: k thanks, i will do it when i have the time

File: lua/entities/gmod_wire_expression2/core/custom/cfc_vector2_NumberOperands.lua (hunk 16 above)
Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_vector2_NumberOperands.lua]: What do you mean by converted to spaces?

Comment (PLally) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_vector2_NumberOperands.lua]: the file is indented with tabs, files should be indented with 4 spaces

Comment (TechnichalDoge) [reply lua/entities/gmod_wire_expression2/core/custom/cfc_vector2_NumberOperands.lua]: thank you

**Comment by TechnichalDoge** [comment]
I split the bill
